pkg/virt-handler/vm.go

```diff
@@ -1128,6 +1128,15 @@ func (d *VirtualMachineController) updateVMIStatus(origVMI *v1.VirtualMachineIns
 		log.Log.Errorf("virt-launcher does not support the Secure Boot setting. Updating VMI %s status to Failed", vmi.Name)
 		vmi.Status.Phase = v1.Failed
 	}
+
+	if vmi.ResourceVersion == "" {
+		vmiCopy, err := d.clientset.VirtualMachineInstance(vmi.ObjectMeta.Namespace).Get(vmi.Name, &metav1.GetOptions{})
+		if err != nil {
+			log.Log.Object(vmi).Infof("Can not get resourceVersion for Updating VMI %s status to Failed, error %v", vmi.Name, err.Error())
+			return err
+		}
+		vmi.ResourceVersion = vmiCopy.ResourceVersion
```

Comment:
Something isn't right about this line.  Kubernetes is responsible for ResourceVersion being up to date so why do we need to update it?  I'm also not sure if we're even allowed to -  `kubectl edit vmi...` to change the ResourceVersion and it doesn't work.

I would look into how the ResourceVersion became `""` in the first place since that might give us a better understanding of the problem.

Reply:
> Kubernetes is responsible for ResourceVersion being up to date so why do we need to update it? 

Kubernetes has MVCC to orchestrate concurrent write to same object. User needs to specifically fetch the latest resrouceVersion in two situations: 
1) The object is concurrently modified by other controllers, like a CAS, the failed update will be enqueue and will never properly processed without fetching latest resourceVersion specifically in user code. 
2) Periodically ReSync happens. Every time resync happens, UpdateFunc will be triggered for each of remaining elements in FIFO queue, there might be some missed or previous failed updates causing resourceVersion out of date, so these objects will not be updated properly and re-enqueue repeatedly.
I think at any case above, we need to add get latest RV for virthanlder. Actually, we have seen this issue frequently happens in virt-handler, see:
```
{"component":"virt-handler","level":"info","msg":"re-enqueuing VirtualMachineInstance ns-03/test-vm","pos":"vm.go:956","reason":"Operation cannot be fulfilled on virtualmachineinstances.kubevirt.io \"test-vm\": the object has been modified; please apply your changes to the latest version and try again","timestamp":"2022-02-10T15:21:28.484737Z"}
```

>  I'm also not sure if we're even allowed to - kubectl edit vmi... to change the ResourceVersion and it doesn't work.

 We don't know what root caused `ResourceVersion became ""` yet, but we have seen the virt-handler has higher chance to lose the resourceVersion to update VMI.

 ```
 {"component":"virt-handler","kind":"","level":"info","msg":"VMI is in phase: Running | Domain status: Running, reason: Unknown","name":"test-vm","namespace":"nns-03","pos":"vm.go:1231","timestamp":"2022-02-14T18:13:07.329329Z","uid":"65b61ceb-315c-446a-a785-1a620eada37a"}
{"component":"virt-handler","kind":"","level":"info","msg":"Signaled graceful shutdown for test-vm","name":"test-vm","namespace":"nns-03","pos":"vm.go:1738","timestamp":"2022-02-14T18:13:07.340591Z","uid":"65b61ceb-315c-446a-a785-1a620eada37a"}
{"component":"virt-handler","kind":"","level":"info","msg":"VMI is in phase: Running | Domain status: Running, reason: Unknown","name":"test-vm","namespace":"nns-03","pos":"vm.go:1231","timestamp":"2022-02-14T18:13:07.375186Z","uid":"65b61ceb-315c-446a-a785-1a620eada37a"}
{"component":"virt-handler","kind":"","level":"error","msg":"Synchronizing the VirtualMachineInstance failed.","name":"test-vm","namespace":"nns-03","pos":"vm.go:1375","reason":"server error. command Shutdown failed: \"context deadline exceeded\"","timestamp":"2022-02-14T18:13:11.379096Z","uid":"65b61ceb-315c-446a-a785-1a620eada37a"}

{"component":"virt-handler","kind":"","level":"error","msg":"Updating the VirtualMachineInstance status failed.","name":"test-vm","namespace":"nns-03","pos":"vm.go:1382","reason":"Operation cannot be fulfilled on virtualmachineinstances.kubevirt.io \"test-vm\": the object has been modified; please apply your changes to the latest version and try again","timestamp":"2022-02-14T18:13:11.422369Z","uid":"65b61ceb-315c-446a-a785-1a620eada37a"}
 ```

Reply:
Maybe we get `ResoruceVersion == ""` because object was deleted from etcd while virt-handler's watch was closed (virt-handler was restarting)?  Here's how [k8s recommends](https://github.com/kubernetes/website/blob/main/content/en/docs/reference/using-api/api-concepts.md#efficient-detection-of-changes) we detect changes:

```
When the requested watch operations fail because the historical version of that resource
is not available, clients must handle the case by recognizing the status code 410 Gone,
clearing their local cache, performing a new get or list operation, and starting the watch
from the resourceVersion that was returned.
```

If we're getting `ResourceVersion == ""` then I think we need to relist.  Looks like there are [some parameters](https://github.com/kubernetes/website/blob/main/content/en/docs/reference/using-api/api-concepts.md#resource-versions) we can pass like `resourceVersionMatch` that might be helpful.  But I don't think we should be modifying `ResourceVersion`:
```
Resource versions are strings that identify the server's internal version of an object.
Resource versions can be used by clients to determine when objects have changed,
or to express data consistency requirements when getting, listing and watching resources.
Resource versions must be treated as opaque by clients and passed unmodified back to the server.
```

Reply:
I don't think VMI is deleted from etcd because I can use **kubectl get vmi** to get this VMI object with latest resourceVersion. 
e.g.
```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  annotations:
    kubevirt.io/latest-observed-api-version: v1alpha3
    kubevirt.io/storage-observed-api-version: v1alpha3
    vm.kubevirt.io/baseDiskName: "...."
    vm.kubevirt.io/deltaDiskName: rootdisk
  creationTimestamp: "2022-04-26T17:58:35Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2022-04-26T20:58:04Z"
  finalizers:
  - foregroundDeleteVirtualMachine
  generation: 283
  labels:
    creationUnixTimestamp: "1650995915"
    performance_class: ga10g_1.br20_2xlarge
    pmApiKey: d27a72a7-a97a-a84a-482a-dfd9ca84339b
    seq: "866406820"
  name: fc2ca756-31a3-4266-99a3-abadf9912b1d
  namespace: np-gs
  resourceVersion: "1416653235"
  uid: 73e3810e-f25a-4580-b566-0a77c9ce1d04
spec:
  ...
```
virt-hanlder should be able to directly fetch this resourceVerson to move forward.